core/new-gui/src/app/workspace/component/property-editor/property-editor.component.ts

```diff
@@ -1,15 +1,234 @@
+    // hide the view first and then make everything null
+    this.displayForm = false;
+
+    this.currentOperatorID = undefined;
+    this.currentOperatorInitialData = undefined;
+    this.currentOperatorSchema = undefined;
+
+  }
+
+  /**
+   * Changes the property editor to use the new operator data.
+   * Sets all the data needed by the json schema form and displays the form.
+   * @param operator
+   */
+  public changePropertyEditor(operator: OperatorPredicate | undefined): void {
+    if (!operator) {
+      throw new Error(`change property editor: operator is undefined`);
+    }
+
+    // set displayForm to false first to hide the view while constructing data
+    this.displayForm = false
+
+    // set the operator data needed
+    this.currentOperatorID = operator.operatorID;
+    this.currentOperatorSchema = this.operatorSchemaList.find(schema => schema.operatorType === operator.operatorType);
+    if (!this.currentOperatorSchema) {
+      throw new Error(`operator schema for operator type ${operator.operatorType} doesn't exist`)
+    }
+    /**
+     * Make a deep copy of the initial property data object.
+     * It's important to make a deep copy. If it's a reference to the operator's property object,
+     *  form change event -> property object change -> input to form change -> form change event
+     *  although the it falls into an infinite loop of tirggering events.
+     * Making a copy prevents that property object change triggers the input to the form changes.
+     *
+     * Although currently other methods also prevent this to happen, it's still good to explicitly make a copy.
+     *  - now the operator property object is immutable, meaning a new property object is construct to replace the old one,
+     *      instead of directly mutating the same object reference
+     *  - now the formChange event handler checks if the new formData is equal to the current operator data,
+     *      which prevents the
+     */
+    this.currentOperatorInitialData = cloneDeep(operator.operatorProperties);
 
-  constructor() { }
+    // set displayForm to true in the end - first initialize all the data then show the view
+    this.displayForm = true;
+  }
+
+  /**
+   * Handles the highlight / unhighlight events.
+   * On highlight -> display the form of the highlighted operator
+   * On unhighlight -> hides the form
+   */
+  public handleHighlightEvents() {
+    this.workflowActionService.getJointGraphWrapper().getJointCellHighlightStream()
+      .filter(value => value.operatorID !== this.currentOperatorID)
+      .map(value => this.workflowActionService.getTexeraGraph().getOperator(value.operatorID))
+      .subscribe(
+        operator => this.changePropertyEditor(operator)
+      );
+
+    this.workflowActionService.getJointGraphWrapper().getJointCellUnhighlightStream()
+      .filter(value => value.operatorID === this.currentOperatorID)
+      .subscribe(value => this.clearPropertyEditor());
+  }
```

Comment:
I see that always in code we do unhighlight and then highlight. So, hopefully clearPropertyEditor() will be invoked first and then changePropertyEditor(). But are we sure that observables' subscriptions are invoked in the order they are called in code?

Reply:
Multiple subscriptions to one observable don't have any guarantee in execution order. 

However in this case it's subscriptions to different observables: the highlight observable and the unhighlight observable. 
So if in the code:
```
unhighlight(); // this will trigger the unhighlight observable -> unhighlight subscriptions will be executed here, even before the function returns
highlight(); // this will trigger the highlight observable -> highlight subscriptions will be executed here
```
So if somewhere else in the code, `unhighlight` and `highlight` is called in proper order, this will also execute properly.